source/lib/api/AST/ASTVisitor.cpp

```diff
@@ -310,26 +325,206 @@ parseParameters(
+        for(const NamedDecl* NND : *NestedParamList)
+        {
+            extinfo.Params.emplace_back(
+                buildTemplateParam(NND));
+        }
+        if(TP->hasDefaultArgument())
+        {
+            extinfo.Default.emplace(getSourceCode(
+                ND, TP->getDefaultArgumentLoc()));
+        }
+    }
+    return info;
+}
+
+void
+ASTVisitor::
+buildTemplateArgs(
+    TemplateInfo& info,
+    ArrayRef<TemplateArgument> args)
+{
+    // TypePrinter generates an internal placeholder name (e.g. type-parameter-0-0)
+    // for template type parameters used as arguments. it also cannonicalizes
+    // types, which we do not want (although, PrintingPolicy has an option to change this).
+    // thus, we use the template arguments as written.
+
+    // KRYSTIAN NOTE: this can probably be changed to select
+    // the argument as written when it is not dependent and is a type.
+    // FIXME: constant folding behavior should be consistent with that of other
+    // constructs, e.g. noexcept specifiers & explicit specifiers
+    const auto& policy = astContext_->getPrintingPolicy();
+    for(const TemplateArgument& arg : args)
+    {
+        std::string arg_str;
+        llvm::raw_string_ostream stream(arg_str);
+        arg.print(policy, stream, false);
+        info.Args.emplace_back(std::move(arg_str));
+    }
+}
+
+void
+ASTVisitor::
+buildTemplateArgs(
+    TemplateInfo& info,
+    ArrayRef<TemplateArgumentLoc> args)
+{
+    // TypePrinter generates an internal placeholder name (e.g. type-parameter-0-0)
+    // for template type parameters used as arguments. it also cannonicalizes
+    // types, which we do not want (although, PrintingPolicy has an option to change this).
+    // thus, we use the template arguments as written.
+
+    // KRYSTIAN NOTE: this can probably be changed to select
+    // the argument as written when it is not dependent and is a type.
+    // FIXME: constant folding behavior should be consistent with that of other
+    // constructs, e.g. noexcept specifiers & explicit specifiers
+    const auto& policy = astContext_->getPrintingPolicy();
+    for(const TemplateArgumentLoc& arg : args)
+    {
+        std::string arg_str;
+        llvm::raw_string_ostream stream(arg_str);
+        arg.getArgument().print(policy, stream, false);
+        info.Args.emplace_back(std::move(arg_str));
+    }
+}
+
 void
 ASTVisitor::
-getTemplateParams(
+parseTemplateArgs(
+    llvm::Optional<TemplateInfo>& Info,
+    const ClassTemplateSpecializationDecl* spec)
+{
+    if(! Info)
+        Info.emplace();
+
+    // USR of the primary template
+    if(ClassTemplateDecl* primary = spec->getSpecializedTemplate())
+    {
+        if(primary->isMemberSpecialization())
+            primary = primary->getInstantiatedFromMemberTemplate();
+        Info->Primary.emplace(getSymbolID(primary));
+    }
+
+    const TypeSourceInfo* tsi = spec->getTypeAsWritten();
+        Assert(tsi && "expected type source information");
```

Comment:
indentation and I don't like this business of putting a string. no strings please, just put that in a comment